keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2397 @@
+before](https://github.com/kubernetes-csi/external-provisioner/issues/86)
+but were essentially [rejected by
+SIG-Storage](https://github.com/kubernetes-csi/external-provisioner/issues/86#issuecomment-465836185)
+because allowing apps to set custom parameters would make apps
+non-portable.
+
+The current volume support also has open issues that affect the
+“volume as resource” approach: Multiple different Pods on a node are
+allowed to use the same
+volume. https://github.com/kubernetes/enhancements/pull/2489 will
+address that, but is still work in progress.  Recovery from a bad node
+selection during delayed binding may get stuck when a Pod has multiple
+volumes because volumes are not getting deleted after a partial
+provisioning. A proposal to fix that needs further work
+(https://github.com/kubernetes/enhancements/pull/1703).  Each “fake”
+CSI driver would have to implement and install a scheduler extender
+because storage capacity tracking only considers volume size as
+criteria for selecting nodes, which is not applicable for custom
+resources.
+
+### Extend volume support
+
+The StorageClass and PersistentVolumeClaim structs could be extended
+to allow custom parameters. Together with an extension of the CSI
+standard that would address the main objection against the previous
+alternative.
+
+However, SIG-Storage and the CSI community would have to agree to this
+kind of reuse and accept that some of the code maintained by them
+becomes more complex because of these new use cases.
+
+### Extend Device Plugins
+
+The device plugins API could be extended to implement some of the
+requirements mentioned in the “Motivation” section of this
+document. There were certain attempts to do it, for example an attempt
+to [add ‘Deallocate’ API call](https://github.com/kubernetes/enhancements/pull/1949) and [pass pod annotations to 'Allocate' API call](https://github.com/kubernetes/kubernetes/pull/61775)
+
+However, most of the requirements couldn’t be satisfied using this
+approach as they would require major incompatible changes in the
+Device Plugins API. For example: partial and optional resource
+allocation couldn’t be done without changing the way resources are
+currently declared on the Pod and Device Plugin level.
+
+Extending the device plugins API to use [Container Device Interface](https://github.com/container-orchestrated-devices/container-device-interface)
+would help address some of the requirements, but not all of them.
+
+NodePrepareResource and NodeUnprepareResource could be added to the Device Plugins API and only get called for
+resource claims.
+
+However, this would mean that
+developers of the device plugins would have to implement mandatory
+API calls (ListAndWatch, Allocate), which could create confusion
+as those calls are meaningless for the Dynamic Resource Allocation
+purposes.
+
+Even worse, existing device plugins would have to implement the new
```

Comment:
I like the idea of adding a more developed alternative option in this KEP, so that final review can see what other options were considered.

I could even imagine that both approaches might make it to shipped alpha code and then we could use real-world experience to decide. Unlikely I suspect but plausible.

Reply:
> Could we introduce a "v2" of the device plugin API with these additional methods instead of adding the methods directly into the current API?

I'm assuming that with device plugin API, you mean the gRPC interface here, not how it is exposed in the PodSpec. Such a v2  (or rather, v2alpha1) wouldn't affect other device plugins, but then kubelet will need to support two versions of that gRPC API, of which the one with the additional methods is unnecessarily complex.

> I think that still be a much smaller change compared to the proposal here and would be much smaller in scope (since it would be limited to kubelet / node level) and would likely not impact other areas mentioned here such as scheduling/core pod api/etc?

But then it fails to achieve most of the goals outlined for this KEP, right?

> I'm curious what requirements would be met with an overhaul of the device plugin API instead...

We have considered this long and hard, but no, some smaller changes to that API won't be sufficient. In a way, this KEP *is* that overhaul of the device plugin API.

Reply:
> I like the idea of adding a more developed alternative option in this KEP, so that final review can see what other options were considered.

I can't think of any solution that adds support for network-attached devices through the device plugin API. I had thought about pretending to have local resources that are backed by network-attached resources, but came to the conclusions that this then would have race conditions because kube-scheduler would potentially "reserve" the same resource instance for two pods at the same time. I could add that to the KEP. I thought it was there, but I don't find it.

One can of course argue that network-attached devices are not worth supporting. But I also can't think of a solution for fitting non-integer parameters into https://pkg.go.dev/k8s.io/api/core/v1#ResourceList (used in PodSpec and NodeStatus) and then having the scheduler do anything meaningful with them.

> I could even imagine that both approaches might make it to shipped alpha code and then we could use real-world experience to decide.

We don't have to :smile: We later learned that Akri implements exactly the approach above for network-attached devices and can indeed run into the problem of pods getting stuck on nodes.

Reply:
Thanks @pohly for the feedback!

> We have considered this long and hard, but no, some smaller changes to that API won't be sufficient. In a way, this KEP is that overhaul of the device plugin API.

That makes sense, but as mentioned in the later part of the KEP, there is opportunity to overhaul the API. I think even if the kubelet had to support two external device plugin APIs, there could be some internal code reuse and it would still be somewhat limited in scope of mostly touching kubelet/node area. I would be interested to understand what specific goals could be accomplish with that approach as compared to solution outlined in this proposal.

> We don't have to 😄 We later learned that Akri implements exactly the approach above for network-attached devices and can indeed run into the problem of pods getting stuck on nodes.

I don't know too much about Akri but it's interesting there is a whole out of tree k8s project which at least on the surface (seems to me) has similar goals here. Do we know more about what limitations they have and how that project's goals differ from the ones reported here (what is "pod stuck on nodes issue")? It may be an interesting case study to include in the KEP since it seems very related.